modules/scripts/startup-script/examples/install_ansible.sh

```diff
@@ -13,41 +13,181 @@
+# newly installed packaged.
+install_pip3_apt() {
+	apt-get update
+	apt-get install -y python3-pip
+}
+
+install_pip3() {
+	if [ -f /etc/centos-release ] || [ -f /etc/redhat-release ] ||
+		[ -f /etc/oracle-release ] || [ -f /etc/system-release ]; then
+		install_pip3_yum
+	elif [ -f /etc/debian_version ] || grep -qi ubuntu /etc/lsb-release 2>/dev/null ||
+		grep -qi ubuntu /etc/os-release 2>/dev/null; then
+		install_pip3_apt
+	else
+		echo "Error: Unsupported Distribution"
+		return 1
+	fi
+}
+
+main() {
+	# Get the python3 executable, or install it if not found
+	get_python_path
+	get_python_major_version "${python_path}"
+	get_python_minor_version "${python_path}"
+	if [ "${python_path}" = "" ] || [ "${python_major_version}" = "2" ] || [ "${python_minor_version}" -lt "${REQ_PYTHON3_VERSION}" ]; then
+		if ! install_python3; then
+			return 1
 		fi
-		/usr/bin/python -m pip install virtualenv
-		/usr/bin/python -m virtualenv /usr/local/toolkit
-		/usr/local/toolkit/bin/python -m pip install wheel
-		/usr/local/toolkit/bin/python -m pip install ansible==2.9.27
-		ln -s /usr/local/toolkit/bin/ansible-playbook /usr/bin/ansible-playbook
-	elif [ -f /etc/debian_version ] || grep -qi ubuntu /etc/lsb-release || grep -qi ubuntu /etc/os-release; then
-		echo 'WARNING: unsupported installation of ansible in debian / ubuntu'
-		apt_wait
-		apt-get update
-		DEBIAN_FRONTEND=noninteractive apt-get install -y ansible
+		get_python_major_version "${python_path}"
+		get_python_minor_version "${python_path}"
 	else
-		echo 'Unsupported distribution'
-		exit 1
+		install_python_deps
 	fi
-fi
+
+	# Install and/or upgrade pip
+	if ! ${python_path} -m pip --version 2>/dev/null; then
+		if ! install_pip3; then
+			return 1
+		fi
+	fi
+	pip_version=$(${python_path} -m pip --version | sed -nr 's/^pip ([0-9]+\.[0-9]+).*$/\1/p')
+	pip_major_version=$(echo "${pip_version}" | cut -d '.' -f 1)
+	echo "pip version: ${pip_version}"
+	if [ "${pip_major_version}" -lt "${REQ_PIP_VERSION}" ]; then
+		${python_path} -m pip install --upgrade pip
+	fi
+
+	# Create pip virtual environment for HPC Toolkit
+	${python_path} -m pip install virtualenv
+	${python_path} -m virtualenv /usr/local/ghpc-venv
+	python_path=/usr/local/ghpc-venv/bin/python3
+
+	# Install ansible
+	ansible_version=""
+	if which ansible-playbook 2>/dev/null; then
+		ansible_version=$(ansible-playbook --version 2>/dev/null | sed -nr 's/^ansible-playbook.*([0-9]+\.[0-9]+\.[0-9]+).*/\1/p')
+		ansible_major_vers=$(echo "${ansible_version}" | cut -d '.' -f 1)
+		ansible_minor_vers=$(echo "${ansible_version}" | cut -d '.' -f 2)
+		ansible_req_major_vers=$(echo "${REQ_ANSIBLE_VERSION}" | cut -d '.' -f 1)
+		ansible_req_minor_vers=$(echo "${REQ_ANSIBLE_VERSION}" | cut -d '.' -f 2)
+	fi
+	if [ -z "${ansible_version}" ] || [ "${ansible_major_vers}" -ne "${ansible_req_major_vers}" ] ||
+		[ "${ansible_minor_vers}" -lt "${ansible_req_minor_vers}" ]; then
+		${python_path} -m pip install ansible==${REQ_ANSIBLE_PIP_VERSION}
```

Comment:
This is making the choice to install 4.10.0 on all platforms (aligning all platforms to the last version compatible with CentOS 7 and Rocky 8). That is a reasonable choice. It might also be reasonable to install whatever `pip` thinks is the best version. Anyhow, I think this is a choice we should all understand.

In particular, I cannot find documentation for 4.10.0 at https://docs.ansible.com/

Reply:
The pip package 4.10.0 will install version 2.11 of ansible-core, which I am able to find listed in the documentation: https://docs.ansible.com/ansible-core/2.11/index.html

It is however the oldest version listed currently. 

I'm inclined to keep the hard version in place for now and investigate installing as `python -m pip install ansible>=4.10.0` when we can run further tests against it (I'll a task for it in the backlog). How do you all feel about that?

Reply:
I like that idea as the best way to make progress.